jax/_src/debugging.py

```diff
@@ -121,14 +121,24 @@ def debug_callback_transpose_rule(*flat_args, callback: Callable[..., Any],
 
 def debug_callback_lowering(ctx, *args, effect, callback, **params):
 
-  if isinstance(ctx.module_context.axis_context,
-                (mlir.SPMDAxisContext, mlir.ShardingContext)):
-    # Apply maximal sharding so pjit only executes the callback on device 0.
+  axis_context = ctx.module_context.axis_context
+  if (isinstance(axis_context, mlir.SPMDAxisContext) and
+        set(axis_context.manual_axes) == set(axis_context.mesh.axis_names)):
+    # If we have fully manual sharding during lowering, that means the JAX
+    # program has per-device semantics, so we run the callback on each device.
+    sharding = xc.OpSharding()
+    sharding.type = xc.OpSharding.Type.MANUAL
+  elif isinstance(axis_context, (mlir.ShardingContext, mlir.SPMDAxisContext)):
+    # If we have fully automatic sharding during lowering, that means the JAX
+    # program has bulk array semantics, so we run the callback with a MAXIMAL
+    # sharding and hence execute it only once on the full logical value).
+    # If we have partially automatic sharding, we do this too... not sure why!
```

Comment:
In my first cut at this logic, it was more like: if fully manual then MANUAL, if fully auto then MAXIMAL, if not those cases but still spmd-ish ERROR, else None. But some tests started failing with that! So I went back to the old behavior: if partially manual / partially automatic we also do MAXIMAL.